File: Sources/LaboLaboEngine/Git/GitModels.swift
import Foundation

/// Parsed snapshot of `git status --porcelain=v2 --branch -z` for one worktree.
public struct GitStatus: Equatable, Sendable {
    public var headSha: String?
    public var branch: String?
    public var upstream: String?
    public var ahead: Int
    public var behind: Int
    public var entries: [GitFileEntry]

    public init(
        headSha: String? = nil,
        branch: String? = nil,
        upstream: String? = nil,
        ahead: Int = 0,
        behind: Int = 0,
        entries: [GitFileEntry] = []
    ) {
        self.headSha = headSha
        self.branch = branch
        self.upstream = upstream
        self.ahead = ahead
        self.behind = behind
        self.entries = entries
    }

    /// `true` when the branch is in a detached-HEAD state (porcelain reports `(detached)`).
    public var isDetached: Bool { branch == "(detached)" }

    /// Files with index-side (staged) changes.
    public var staged: [GitFileEntry] {
        entries.filter { $0.isStaged }
    }

    /// Files with worktree-side (unstaged) changes.
    public var unstaged: [GitFileEntry] {
        entries.filter { $0.isUnstaged }
    }

    /// Untracked paths.
    public var untracked: [GitFileEntry] {
        entries.filter { $0.kind == .untracked }
    }

    /// Conflicted (unmerged) paths.
    public var conflicted: [GitFileEntry] {
        entries.filter { $0.kind == .unmerged }
    }

    public var isDirty: Bool {
        entries.contains { $0.kind != .ignored }
    }
}

/// A single changed-path record from porcelain v2.
public struct GitFileEntry: Equatable, Sendable {
    public enum Kind: Equatable, Sendable {
        case ordinary
        case renamedOrCopied
        case unmerged
        case untracked
        case ignored
    }

    /// Status code for one side of an `XY` pair. `.unmodified` is porcelain's `.`.
    public enum Change: Character, Equatable, Sendable {
        case unmodified = "."
        case modified = "M"
        case typeChanged = "T"
        case added = "A"
        case deleted = "D"
        case renamed = "R"
        case copied = "C"
        case updatedButUnmerged = "U"

        public init(porcelain: Character) {
            self = Change(rawValue: porcelain) ?? .unmodified
        }
    }

    public var kind: Kind
    /// Index (staged) side of the `XY` field.
    public var index: Change
    /// Worktree (unstaged) side of the `XY` field.
    public var worktree: Change
    public var path: String
    /// Original path for rename/copy entries.
    public var originalPath: String?
    /// Similarity/score for rename/copy entries (0–100), if present.
    public var score: Int?

    public init(
        kind: Kind,
        index: Change = .unmodified,
        worktree: Change = .unmodified,
        path: String,
        originalPath: String? = nil,
        score: Int? = nil
    ) {
        self.kind = kind
        self.index = index
        self.worktree = worktree
        self.path = path
        self.originalPath = originalPath
        self.score = score
    }

    public var isStaged: Bool {
        switch kind {
        case .ordinary, .renamedOrCopied:
            return index != .unmodified
        default:
            return false
        }
    }

    public var isUnstaged: Bool {
        switch kind {
        case .ordinary, .renamedOrCopied:
            return worktree != .unmodified
        case .unmerged:
            return true
        default:
            return false
        }
    }
}

File: rust/crates/labolabo-core/src/git_models.rs
//! Faithful port of `Sources/LaboLaboEngine/Git/GitModels.swift`.
//!
//! These are plain data types plus the small set of derived (computed)
//! properties the Swift version exposes. Behavior — including which entries
//! count as "staged"/"unstaged"/"dirty" — must match the Swift source
//! exactly; see the doc comments below for the specific rules carried over.

/// Parsed snapshot of `git status --porcelain=v2 --branch -z` for one worktree.
#[derive(Debug, Clone, PartialEq, Eq, Default)]
pub struct GitStatus {
    pub head_sha: Option<String>,
    pub branch: Option<String>,
    pub upstream: Option<String>,
    pub ahead: i64,
    pub behind: i64,
    pub entries: Vec<GitFileEntry>,
}

impl GitStatus {
    /// `true` when the branch is in a detached-HEAD state (porcelain reports `(detached)`).
    pub fn is_detached(&self) -> bool {
        self.branch.as_deref() == Some("(detached)")
    }

    /// Files with index-side (staged) changes.
    pub fn staged(&self) -> Vec<&GitFileEntry> {
        self.entries.iter().filter(|e| e.is_staged()).collect()
    }

    /// Files with worktree-side (unstaged) changes.
    pub fn unstaged(&self) -> Vec<&GitFileEntry> {
        self.entries.iter().filter(|e| e.is_unstaged()).collect()
    }

    /// Untracked paths.
    pub fn untracked(&self) -> Vec<&GitFileEntry> {
        self.entries
            .iter()
            .filter(|e| e.kind == Kind::Untracked)
            .collect()
    }

    /// Conflicted (unmerged) paths.
    pub fn conflicted(&self) -> Vec<&GitFileEntry> {
        self.entries
            .iter()
            .filter(|e| e.kind == Kind::Unmerged)
            .collect()
    }

    /// NOTE: this intentionally counts *untracked* entries as "dirty" too —
    /// it is `true` whenever any entry's kind is not `.ignored`, not just
    /// staged/unstaged ones. This mirrors the Swift `isDirty` exactly.
    pub fn is_dirty(&self) -> bool {
        self.entries.iter().any(|e| e.kind != Kind::Ignored)
    }
}

/// A single changed-path record from porcelain v2.
#[derive(Debug, Clone, PartialEq, Eq)]
pub struct GitFileEntry {
    pub kind: Kind,
    /// Index (staged) side of the `XY` field.
    pub index: Change,
    /// Worktree (unstaged) side of the `XY` field.
    pub worktree: Change,
    pub path: String,
    /// Original path for rename/copy entries.
    pub original_path: Option<String>,
    /// Similarity/score for rename/copy entries (0-100), if present.
    pub score: Option<i64>,
}

impl GitFileEntry {
    pub fn new(kind: Kind, path: impl Into<String>) -> Self {
        Self {
            kind,
            index: Change::Unmodified,
            worktree: Change::Unmodified,
            path: path.into(),
            original_path: None,
            score: None,
        }
    }

    /// Ordinary/renamed-or-copied entries are staged when their index side
    /// changed. Unmerged/untracked/ignored entries are never "staged"
    /// (unmerged conflicts show up as unstaged instead — see `is_unstaged`).
    pub fn is_staged(&self) -> bool {
        match self.kind {
            Kind::Ordinary | Kind::RenamedOrCopied => self.index != Change::Unmodified,
            _ => false,
        }
    }

    /// Ordinary/renamed-or-copied entries are unstaged when their worktree
    /// side changed. Unmerged entries are *always* unstaged (regardless of
    /// their XY code), matching the Swift source's unconditional `true`.
    pub fn is_unstaged(&self) -> bool {
        match self.kind {
            Kind::Ordinary | Kind::RenamedOrCopied => self.worktree != Change::Unmodified,
            Kind::Unmerged => true,
            _ => false,
        }
    }
}

#[derive(Debug, Clone, Copy, PartialEq, Eq)]
pub enum Kind {
    Ordinary,
    RenamedOrCopied,
    Unmerged,
    Untracked,
    Ignored,
}

/// Status code for one side of an `XY` pair. `Unmodified` is porcelain's `.`.
#[derive(Debug, Clone, Copy, PartialEq, Eq)]
pub enum Change {
    Unmodified,
    Modified,
    TypeChanged,
    Added,
    Deleted,
    Renamed,
    Copied,
    UpdatedButUnmerged,
}

impl Change {
    /// Any character not among the recognised porcelain codes falls back to
    /// `Unmodified`, mirroring the Swift `Change(porcelain:)` initializer's
    /// `?? .unmodified` fallback (no error, no panic).
    pub fn from_porcelain(c: char) -> Change {
        match c {
            '.' => Change::Unmodified,
            'M' => Change::Modified,
            'T' => Change::TypeChanged,
            'A' => Change::Added,
            'D' => Change::Deleted,
            'R' => Change::Renamed,
            'C' => Change::Copied,
            'U' => Change::UpdatedButUnmerged,
            _ => Change::Unmodified,
        }
    }

    /// Inverse of `from_porcelain`; the single-character porcelain code.
    pub fn to_porcelain(self) -> char {
        match self {
            Change::Unmodified => '.',
            Change::Modified => 'M',
            Change::TypeChanged => 'T',
            Change::Added => 'A',
            Change::Deleted => 'D',
            Change::Renamed => 'R',
            Change::Copied => 'C',
            Change::UpdatedButUnmerged => 'U',
        }
    }
}
